### PR TITLE
Ingame Atmos Reaction Guide

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
~~https://imgur.com/a/pMMEi4i~~
https://imgur.com/a/xCrIcz4

Title, really.
Adds an ingame guide to atmos. Currently hooked to the atmos monitors, analyzer, and the tablet app.

~~Lots of reaction data not implemented yet, banking on the cleanup to get merged first, so drafting.~~
Done, all reactions in. Haven't double checked them though.

Code is pretty much finished, feel free to take a look. Ill probably retidy them while adding stuffs so no rush.

~~Might add a reaction handbook obj later~~ Implemented in analyzer.

Dotted tooltip idea shamelessly stolen from preferences.

Lots of the diffs are from breaking the sensor file up, dont worry about it.

## Why It's Good For The Game
Less need to open the wiki in another page I guess.

## Changelog
:cl:
add: Added an ic atmos reaction guide. Available in your atmos control consoles/monitors, ntosatmos app, and analyzer.
code: Some changes to how gas canister descriptions are generated. 
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/reaction_factors.dm
Comment (Ghilker) [inline code/modules/atmospherics/gasmixtures/reaction_factors.dm:3]: i know is a pain but i think the numbers and amounts should not be hardcoded in but they should be calculated in some way so that if you change the reaction these would automatically import the new values

Comment (vincentiusvin) [reply code/modules/atmospherics/gasmixtures/reaction_factors.dm:3]: I agree. Which is why Im waiting for temp's reaction cleanup pr to get in first.

**Comment by Mothblocks** [comment]
I don't like the underlines, they don't make me think of links. They should just be normal buttons.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:127]: Why are these `{'0.5em'}` instead of just `'0.5em'`?

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:119]: Why `toLocaleLowerCase` instead of `toLowerCase`? I10n shouldn't matter here.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:138]: Just use `Object.entries` instead of `Object.keys` and then getting value.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:151]: Why shared state?

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:156]: Why shared state?

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:149]: Since you're using TypeScript, you should either not include data, or type it with `useBackend<T>(context)`.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:164]: Unused?

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:205]: This ternary is really hard to read, can you at least put it in more parentheses?

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:216]: Weird parentheses?

File: tgui/packages/tgui/interfaces/common/GasmixParser.tsx
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/common/GasmixParser.tsx:10]:
```suggestion
  gases: [string, string, number][];
```

File: tgui/packages/tgui/styles/main.scss
Comment (Mothblocks) [inline tgui/packages/tgui/styles/main.scss:66]: ABC order.

File: tgui/packages/tgui/styles/interfaces/AtmosHandbook.scss
Comment (Mothblocks) [inline tgui/packages/tgui/styles/interfaces/AtmosHandbook.scss:2]: style doesn't like it when I do it but if your SCSS is this small consider just using `style` 😶

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:48]: Needed to give the input toggle a state. Should i use one of the state hooks?

File: tgui/packages/tgui/interfaces/AtmosControlConsole.js
Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/AtmosControlConsole.js:10]: no real reason to be honest, i'll switch it over to local

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:119]: i have no idea what localelowercase it, must've slipped when it autosuggested.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx]: yes, damn.

File: code/__HELPERS/atmospherics.dm
Comment (vincentiusvin) [reply code/__HELPERS/atmospherics.dm]: Assuming you want them spaced instead. Let me know if you wanted the space indentation to be removed.

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (vincentiusvin) [reply tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx]: prettier did it 👉

File: tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/common/AtmosHandbook.tsx:48]: Yeah, just `useLocalState`

**Comment by vincentiusvin** [comment]
image
not the biggest fan of how it looks but yeah it gets the point across.

**Comment by vincentiusvin** [comment]
Aight, update:
https://imgur.com/a/xCrIcz4

The analyzer caches your last gasmix data, upon which you can read more about the gas by right clicking your analyzer.
The component makes it a bit more annoying to implement, but it's possible. I personally prefer the interaction staying on attack_self, but I'm open if a maint wants it to be in the component. It's removed for now but I can bring it back easily.

**Comment by CourageTheCat** [comment]
thank god this will be so helpful with new toxins

File: code/modules/atmospherics/gasmixtures/reactions.dm
Comment (vincentiusvin) [inline code/modules/atmospherics/gasmixtures/reactions.dm:544]: Believe I should highlight this. Rewrote this a bit because why not. No changes expected though, but still would appreciate a second pair of eyes to recheck. More info can be found at the commit d3079b7db5838604f68c548f989409f5625b13a3.

File: code/modules/atmospherics/gasmixtures/reactions.dm
Comment (vincentiusvin) [inline code/modules/atmospherics/gasmixtures/reactions.dm:77]: Hope this is fine.

File: code/modules/atmospherics/gasmixtures/reactions.dm
Comment (LemonInTheDark) [reply code/modules/atmospherics/gasmixtures/reactions.dm:552]: ❤️

**Comment by vincentiusvin** [comment]
Gonna leave the conflict there for a while. WIll need to rebase after the linked pr is merged or closed anyway.

**Comment by LemonInTheDark** [comment]
ready?

**Comment by vincentiusvin** [comment]
Ready!

**Pull request merged**
